File: fibonacci_algos/fibonacci_sum_last_digit/fibonacci_sum_last_digit.py
# Uses python3
import sys

def fibonacci_sum_naive(n):
    if n <= 1:
        return n

    previous = 0
    current  = 1
    sum      = 1

    for _ in range(n - 1):
        previous, current = current, previous + current
        sum += current

    return sum % 10

def fibonacci_sum_last_digit_fast(n):
    if n <= 1:
        return n

    sum = 0

    max_i = 0

    fibonacci_mod = [0, 1]

    for i in range (2, n+1):
        fibonacci_mod.append((fibonacci_mod[i-1] +fibonacci_mod[i-2])%10)
        
        max_i = i

        if fibonacci_mod[i-1] == 0 and fibonacci_mod[i] == 1:
            break


    periodiclen = len(fibonacci_mod)

    if max_i < n and max_i != 0:
        periodiclen = periodiclen - 2

    n1 = n // periodiclen
    remain = n % periodiclen

    #print('n1=', n1, ' remain=', remain)

    for i in range(periodiclen):
        sum = ( sum + fibonacci_mod[i] ) % 10

    sum = (sum * n1) % 10

    for i in range(remain + 1):
        sum = ( sum + fibonacci_mod[i] ) % 10

    return int(sum)


if __name__ == '__main__':
    #input = sys.stdin.read()
    n = int(input())
    print(fibonacci_sum_last_digit_fast(n))

File: fibonacci_algos/fibonacci_sum_last_digit/FibonacciSumLastDigit.java
import java.util.*;

public class FibonacciSumLastDigit {
    private static long getFibonacciSumNaive(long n) {
        if (n <= 1)
            return n;

        long previous = 0;
        long current  = 1;
        long sum      = 1;

        for (long i = 0; i < n - 1; ++i) {
            long tmp_previous = previous;
            previous = current;
            current = tmp_previous + current;
            sum += current;
        }

        return sum % 10;
    }

    private static int getFibonacciSumLastDigit(long n) {
        if ( n<=1 ) 
            return (int)n;

        ArrayList<Integer> fibonaccimod = new ArrayList<>();

        fibonaccimod.add(0);
        fibonaccimod.add(1);

        int sum = 0;

        long maxI = 0;

        for (int i = 2; i <= n; i++) {
            fibonaccimod.add((fibonaccimod.get(i - 1) + fibonaccimod.get(i - 2))%10);
            //System.out.println("i=" + i + " fib=" + fibonaccimod.get(i));
            //sum = sum +fibonacci.get(i);

            maxI = i;

            if(fibonaccimod.get(i-1) == 0 && fibonaccimod.get(i) == 1) {
                break;
            }
        }

        int periodiclen =  fibonaccimod.size();

        if(maxI < n && maxI != 0) 
            periodiclen -= 2;

        long n1 =  n / periodiclen;

        long remain =  n % periodiclen;

        //System.out.println("periodiclen=" + periodiclen + " n1=" + n1 + " remain=" + remain);

        for (int i = 0; i < periodiclen; i++)
            sum = (sum + fibonaccimod.get(i)) % 10;

        sum = (int) (n1 * sum) % 10;

        for (int i = 0; i <= remain; i++)
            sum = (sum + fibonaccimod.get(i))%10;

        //System.out.println("sum=" + sum);
        return sum;

    }
    
    public static void main(String[] args) {
        //Stress testing

        /*while(true) {
            //Generate n
            Random rand = new Random();
            long n = Math.abs(rand.nextLong()) % 100000000000000L;
            System.out.println("n=" + n);
            System.out.println(getFibonacciSumLastDigit(n));

        }*/

        Scanner scanner = new Scanner(System.in);
        long n = scanner.nextLong();
        int s = getFibonacciSumLastDigit(n);
        System.out.println(s);
    }
}

